Automattic/newspack-ads
# feat(header-bidding): gam orders and line items push

# Header Bidding GAM Integration

This PR implements ad ops requirements for the integration of Prebid header bidding and Google Ad Manager. It follows the [Prebid documentation](https://docs.prebid.org/adops/step-by-step.html) and is inspired by [dfp-prebid-setup](https://github.com/kmjennison/dfp-prebid-setup).

![image](https://user-images.githubusercontent.com/820752/149026845-49cb7211-7e32-4ce0-9451-c905e6246c36.png)

## Structure

### Initial setup
The following objects are created once and should be universal to any header bidding customization.

#### Advertiser
A "company" of type advertiser called "Newspack Header Bidding". It parents every object created in this setup.

#### Targeting keys
The `hb_pb` key, used to segment every existing price bucket in the setup. This is the only targeting key created on setup because its values must be targeted on line item creation.

#### Creatives
10 creatives containing the [Prebid Universal Creative](https://docs.prebid.org/overview/prebid-universal-creative.html). This amount is arbitrary and intended to match at least the number of ad units displayed in a single page.

### Order setup
The order setup is intended to be per price granularity. This implementation restricts the price granularity to our custom "dense" price buckets, but it's built to allow customization if it's decided to change in the future.
 
#### Order
The GAM order that houses the line items and creatives associations.

#### Line items
One line item per price bucket, limited to 450 per order. The custom "dense" granularity generates 449 price buckets.

#### Line Items to Creatives associations (LICAs)
Associate each creative to each order line item, creating LI*C associations. Should sum 4490 LICAs according to our current parameters.

## Wizard

The **initial setup** is performed silently on header bidding activation, whereas the **order setup** requires user interaction through a passive but step-based configuration to prevent timeouts and better handle errors on each step.

![order-progress](https://user-images.githubusercontent.com/820752/149170850-6f4fe802-21b5-4c4b-8c8c-e2629a8297ae.gif)

The most time-consuming steps are the LICAs creations, performed in batches of 500 per request which requires 9 steps for our current setup.

The complete setup should take up to 15 minutes to finish.

Closes #253 

## How to test

 1. Check out this branch and https://github.com/Automattic/newspack-plugin/pull/1378, which provides the JS hooks for our wizard
 2. Make sure you have the experimental header bidding setup properly configured as described in #231 
 3. Without a GAM connection, visit the Advertising -> Settings page and confirm the message "Credentials not found" and the inability to configure the integration
 4. Setup your GAM connection either through a Service Account or OAuth
 5. Visit the Advertising -> Settings page and confirm you can configure the order
 6. If you are using the Newspack test network, create the order with a different name, otherwise it might fail with a "not unique name" error.
 7. Change the name and click on Create Order
 8. Wait until it's finished without leaving the page:
    1. Click outside of the Modal and confirm it does not leave
    2. Attempt to refresh the page and confirm the `onbeforeunload` confirmation. Cancel so you can finish the creation.
